### ISSUE TITLE
Bitwise operators and logical operators

### ISSUE THREAD

**Issue by fdwr** [webmachinelearning/webnn#496]
(separated from my comment here https://github.com/webmachinelearning/webnn/pull/478/files#r1406997415)

Looking ahead, WebNN will eventually want to implement the bitwise operators for completeness as the more mature ML libraries have them, and as the "bitwise" ops are more fundamental than the "logical" ones (the logical ones are a limited subset that can be implemented in terms of the bitwise ones). See the table below for "bitwise" operators and "logical" counterparts. In the [transformers support thread](https://github.com/webmachinelearning/webnn/issues/375#issuecomment-1674224992), I only proposed the minimum needed operators for the pertinent models (just `logicalNot`) rather than the full collection of them, but it's prudent to look holistically when naming them.

Shortening the names (e.g. trimming `bitwiseAnd` or `logicalAnd` -> just `and`) leads to ambiguity of *which* of the two "and"'s is meant. That's probably why most libraries spell them out to disambiguate. One ML library, ONNX, chose the shorter form like `And` to mean a `LogicalAnd`, but it just as easily could have arbitrarily blessed `BitwiseAnd` instead (and arguably that would have made sense because BitwiseAnd is the generic op). So, to reduce ambiguity and increase consistency with library precedent, I think we shouldn't shorten the name "logicalNot" to just "not".

| Bitwise ops | Logical ops |
|--------------|-------------|
| torch.bitwise_and | torch.logical_and |
| torch.bitwise_or | torch.logical_or |
| torch.bitwise_xor | torch.logical_xor |
| torch.bitwise_not | torch.logical_not |
| torch.bitwise_left_shift | - |
| torch.bitwise_right_shift | - |
| tf.bitwise.bitwise_and | tf.math.logical_and |
| tf.bitwise.bitwise_or | tf.math.logical_or |
| tf.bitwise.bitwise_xor | tf.math.logical_xor |
| tf.bitwise.invert | tf.math.logical_not |
| tf.bitwise.left_shift | - |
| tf.bitwise.right_shift | - |
| numpy.bitwise_and | numpy.logical_and |
| numpy.bitwise_or | numpy.logical_or |
| numpy.bitwise_xor | numpy.logical_xor |
| numpy.invert | numpy.logical_not |
| numpy.left_shift | - |
| numpy.right_shift | - |
| tosa.BITWISE_AND | tosa.LOGICAL_AND |
| tosa.BITWISE_OR | tosa.LOGICAL_OR |
| tosa.BITWISE_XOR | tosa.LOGICAL_XOR |
| [tosa.BITWISE_NOT](https://www.mlplatform.org/tosa/tosa_spec.html#_bitwise_not) | [tosa.LOGICAL_NOT](https://www.mlplatform.org/tosa/tosa_spec.html#_logical_not) |
| tosa.LOGICAL_LEFT_SHIFT | - |
| tosa.LOGICAL_RIGHT_SHIFT | - |
| DML_ELEMENT_WISE_BIT_AND | DML_ELEMENT_WISE_LOGICAL_AND |
| DML_ELEMENT_WISE_BIT_OR | DML_ELEMENT_WISE_LOGICAL_OR |
| DML_ELEMENT_WISE_BIT_XOR | DML_ELEMENT_WISE_LOGICAL_XOR |
| DML_ELEMENT_WISE_BIT_NOT | DML_ELEMENT_WISE_LOGICAL_NOT |
| DML_ELEMENT_WISE_BIT_SHIFT_LEFT | - |
| DML_ELEMENT_WISE_BIT_SHIFT_RIGHT | - |
| ONNX BitwiseAnd | ONNX And |
| ONNX BitwiseOr| ONNX Or |
| ONNX BitwiseXor | ONNX Xor |
| ONNX BitwiseNot | ONNX Not |
| - | CoreML LogicalAndLayerParams |
| - | CoreML LogicalOrLayerParams |
| - | CoreML LogicalXorLayerParams |
| - | CoreML LogicalNotLayerParams |

**Comment by shiyi9801** [comment]
Currently there is still an inconsistency of naming between the [spec (`not`)](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-logical) and the [implementation (`logicalNot`)](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/webnn/ml_graph_builder.idl;l=254), should we finally decide the names so we can either change the spec or the implementation to avoid confusion? @huningxin @fdwr

**Comment by inexorabletash** [comment]
I'm supportive of explicit `logical` / `bitwise` prefixes and thus updating spec to `logicalNot()` to match the implementation SGTM.

**Comment by inexorabletash** [comment]
Don't we want this open until we add bitwise and logical ops?

**Comment by fdwr** [comment]
> Don't we want this open until we add bitwise and logical ops?

@inexorabletash Aah, *this* issue was really about the name change, and the large table was just to justify that name change. I'll open a new issue when it comes time to actually introduce the additional bitwise/logical WebNN ops, following the new operator template with the justification and relevant models. ⏳

**Comment by fdwr** [comment]
p.s. The time's coming closer to consider adding these, but the ❌'s above in CoreML might stymie adding a few (I'm having a tough time imagining how to emulate a bitwiseAnd without a true bitwiseAnd, short of a horribly long graph of operators), but I've forwarded the feature request to Mike Wyrzykowski at Apple to fill that gap someday.

**Comment by mwyrzykowski** [comment]
@fdwr I checked with CoreML, bitwise ops is not supported at the ANE compiler -> assembly level. So the request to add it, while possible in the sense that anything is possible, is a bit more challenging / will take longer, it is not something CoreML can simply add.

If we exposed bitwise ops, any implementation running on Apple's platforms would fall off the NPU path which may be surprising to the developer. 

Alternatively, maybe we could support bitwise ops as an extension the developer must opt-in to so they understand it may not be supported across all devices.
